### DISCUSSION TITLE
No option for OKTA on logon screen

### DISCUSSION THREAD

**Discussion by benek984** [netbox-community/netbox#12097] [Help Wanted!]
### NetBox version

v3.4.6

### Python version

3.8

### Steps to Reproduce

I have setup app in OKTA so I got all relevant information for configuratio.py

Added below line to configuration.py
````
REMOTE_AUTH_BACKEND = 'social_core.backends.okta_openidconnect.OktaOpenIdConnect'
SOCIAL_AUTH_OKTA_OPENIDCONNECT_KEY = 'xxxxx'
SOCIAL_AUTH_OKTA_OPENIDCONNECT_SECRET = 'xxxxx'
SOCIAL_AUTH_OKTA_OPENIDCONNECT_API_URL = 'https://xxxx.okta.com/oauth2/'
````
where xxxx my proper values were specyfied

### Expected Behavior

Expecting OKTA logon options on the logon screen.

### Observed Behavior

NO OKTA LOGON OPTION ON THE LOGON SCREE

**Comment by benek984** [comment]
When trying access app from OKTA dashboard this appears: 

![image](https://user-images.githubusercontent.com/62546477/228843826-0f111849-379a-46c9-a80f-8ffd5438e43f.png)

**Reply by Dragnell87** [reply]
I am in the same problem with keycloack, did you find the solution ?
![image (3)](https://user-images.githubusercontent.com/8634449/230050403-506fa8c1-9023-48c8-b5da-cdac986e19f5.png)

**Reply by candlerb** [reply]
What did you put in configuration.py?

I did get Netbox to work with Keycloak last year, but I had to contribute a [fix](https://github.com/python-social-auth/social-core/pull/670) to python-social-auth to make it work. This was merged, so current versions of PSA should have it.

According to my notes, this is what I did (but I won't guarantee they'll work today, as I'm now using Hashicorp Vault instead of Keycloak, and mod_auth_openidc instead of PSA)

Keycloak doesn't/didn't support OIDC discovery so you have to give the exact endpoint URLs.

```
REMOTE_AUTH_ENABLED = True
REMOTE_AUTH_BACKEND = 'social_core.backends.keycloak.KeycloakOAuth2'
REMOTE_AUTH_AUTO_CREATE_USER = True
REMOTE_AUTH_DEFAULT_GROUPS = []
REMOTE_AUTH_DEFAULT_PERMISSIONS = {}
SOCIAL_AUTH_KEYCLOAK_KEY = 'netbox-test'
SOCIAL_AUTH_KEYCLOAK_SECRET = 'XXXXXXXX'
SOCIAL_AUTH_KEYCLOAK_PUBLIC_KEY = 'XXXXXXXXXXXVERYLONGXXXXXXXXXXXXX'
SOCIAL_AUTH_KEYCLOAK_AUTHORIZATION_URL = 'https://keycloak.example.net/realms/example.net/protocol/openid-connect/auth'
SOCIAL_AUTH_KEYCLOAK_ACCESS_TOKEN_URL = 'https://keycloak.example.net/realms/example.net/protocol/openid-connect/token'
```

Also, the [documentation](https://python-social-auth.readthedocs.io/en/latest/backends/keycloak.html) says you need to create a mapper for "audience", and this true - Netbox will reject the authentication unless you do this.
![image](https://user-images.githubusercontent.com/44789/230057026-c69d4e33-d738-4839-9591-73fd805d4141.png)

Finally: with python-social-auth there is (or was) no group/role syncing to Keycloak - groups are maintained entirely within Netbox itself.

**Reply by candlerb** [reply]
I do still have the Keycloak server instance. You can see below what needs to go in the OIDC client configuration.

![image](https://user-images.githubusercontent.com/44789/230059228-bc6d58e3-9314-4053-b247-070cf1309366.png)
![image](https://user-images.githubusercontent.com/44789/230059122-a4345d27-5e30-465a-9520-1bdbabdbeab6.png)

**Comment by Dragnell87** [comment]
Thanks i will look into this,  my current info is:
environment:
    - REMOTE_AUTH_ENABLE=True
    - REMOTE_AUTH_BACKEND=social_core.backends.keycloak.KeycloakOAuth2
    - REMOTE_AUTH_HEADER=HTTP_REMOTE_USER
    - REMOTE_AUTH_AUTO_CREATE_USER=True
    - REMOTE_AUTH_GROUP_HEADER=HTTP_X_FORWARDED_GROUPS
    - REMOTE_AUTH_GROUP_SYNC_ENABLED=True
    - REMOTE_AUTH_SUPERUSER_GROUPS=['netbox-admin']
    - SOCIAL_AUTH_KEYCLOAK_KEY=netbox
    - SOCIAL_AUTH_KEYCLOAK_SECRET=secret
    - SOCIAL_AUTH_KEYCLOAK_PUBLIC_KEY=MIIBILONG
    - SOCIAL_AUTH_KEYCLOAK_AUTHORIZATION_URL=https://infra-auth.local/realms/local/protocol/openid-connect/auth
    - SOCIAL_AUTH_KEYCLOAK_ACCESS_TOKEN_URL=https://infra-auth.local/realms/local/protocol/openid-connect/token

**Reply by candlerb** [reply]
Based on my experience, the redirect_uri configured in Keycloak for this client should be `/oauth/complete/keycloak/` not `/oauth/login/keycloak/`

Aside: REMOTE_AUTH_HEADER and REMOTE_AUTH_GROUP_HEADER won't be used at all with social_core.

**Reply by Dragnell87** [reply]
You mean in keycloack ? It is, no idea where netbox is getting that url from.
![Captura desde 2023-04-05 13-27-18](https://user-images.githubusercontent.com/8634449/230067499-d62637c0-41a2-470a-9518-ab47e61674cc.png)

**Reply by candlerb** [reply]
I suggest you open your browser's developer console and do a login: it should capture all the requests between client-Netbox and client-Keycloak. You should be able to see if the user is being redirected to Keycloak correctly, and what redirect_uri parameter is given, and what URI it gets redirected back to.

**Reply by Dragnell87** [reply]
![Captura desde 2023-04-04 16-05-05](https://user-images.githubusercontent.com/8634449/230153496-9575c1c5-01ea-414b-9f44-d602a0e93ccf.png)
Din´t find anything usefull i belive, no clue from where the redirect_uri is coming from. The keycloack.py need to be edited? 
Are this values ok? or need to be changed ?
    name = "keycloak"
    ID_KEY = "username"

Sorry to bother you. Thanks for all the time.

**Reply by candlerb** [reply]
> The keycloack.py _[sic]_ need to be edited?

Which file do you mean? In social_core? No, you shouldn't touch any library code. The settings in your configuration.py override this.

You are showing that Netbox is fetching `/auth/login/keycloak/?next=%2Foauth%2Flogin%2Fkeycloak%2FNone`. The "next" URL is clearly wrong: it should be the page you were trying to access *prior* to being forced to the login page.

But what's more interesting is the steps which involve communication with the Keycloak server, and those are missing from your screenshot.

What happens if you go to the root page, https://netbox.example.net/ or whatever? Maybe the Keycloak login already took place, and you'll see the home page successfully.  If it redirects you back to the login page, then clicking on "Login with Keycloak" button *should* redirect you to the Keycloak server, i.e. https://infra-auth.local/..., and therefore your browser should show these additional accesses.

**Reply by candlerb** [reply]
> Thanks i will look into this, my current info is:
> ```
> environment:
> ...
> - SOCIAL_AUTH_KEYCLOAK_AUTHORIZATION_URL=https://infra-auth.local/realms/local/protocol/openid-connect/auth
> - SOCIAL_AUTH_KEYCLOAK_ACCESS_TOKEN_URL=https://infra-auth.local/realms/local/protocol/openid-connect/token
> ```

Aha, you mention "environment" - are you running Netbox in a Docker container by any chance?

If you are, all bets are off.  Those settings should go directly in configuration.py, with quotation marks around the string values. netbox-docker only supports a subset of settings via environment variables: see its [configuration.py](https://github.com/netbox-community/netbox-docker/blob/release/configuration/configuration.py)

However it is possible to create an additional file `extra.py` which you mount inside the container, containing the necessary settings. See [here](https://github.com/netbox-community/netbox-docker/wiki/Configuration#custom-configuration-files).

**Reply by Dragnell87** [reply]
Yep, that was IT. Let me invite you to a beer!
